tensornetwork/backends/jax/jitted_functions.py

```diff
@@ -58,84 +62,86 @@ def jax_lanczos(matvec: Callable, arguments: List, init: jax.ShapedArray,
-    krylov_vecs = jax.numpy.zeros((ncv + 1, numel), dtype=init.dtype)
-    norms = jax.numpy.zeros(ncv, dtype=init.dtype)
-    diag_elems = jax.numpy.zeros(ncv, dtype=init.dtype)
-
-    norms = jax.ops.index_update(norms, jax.ops.index[0], 1.0)
-
-    norms_dtype = jax.numpy.real(jax.numpy.empty((0, 0),
-                                                 dtype=init.dtype)).dtype
-    initvals = [
-        init, krylov_vecs, norms, diag_elems, matvec, arguments,
-        norms_dtype.type(1.0), landelta, 1, ncv
-    ]
-    output = jax.lax.while_loop(cond_fun, body_lanczos, initvals)
-    final_state, krylov_vecs, norms, diags, _, _, _, _, it, _ = output
-    krylov_vecs = jax.ops.index_update(krylov_vecs, jax.ops.index[it, :],
-                                       jax.numpy.ravel(final_state))
-
-    A_tridiag = jax.numpy.diag(diags) + jax.numpy.diag(
-        norms[1:], 1) + jax.numpy.diag(jax.numpy.conj(norms[1:]), -1)
+      betas, i = vals[-2], vals[-1]
+      norm = betas[i - 1]
+      return jax.lax.cond(i <= ncv, lambda x: x[0] > x[1], lambda x: False,
+                          [norm, landelta])
+
+    numel = np.prod(shape).astype(np.int32)
+    # note: ncv + 2 because the first vector is all zeros, and the
+    # last is the unnormalized residual.
+    krylov_vecs = jax.numpy.zeros((ncv + 2, numel), dtype=dtype)
+    # NOTE (mganahl): initial state is normalized inside the loop
+    krylov_vecs = krylov_vecs.at[1, :].set(jax.numpy.ravel(init))
+
+    # betas are the upper and lower diagonal elements
+    # of the projected linear operator
+    # the first two beta-values can be discarded
+    # set betas[0] to 1.0 for initialization of loop
+    # betas[2] is set to the norm of the initial state.
+    betas = jax.numpy.zeros(ncv + 1, dtype=dtype)
+    betas = betas.at[0].set(1.0)
+    # diagonal elements of the projected linear operator
+    alphas = jax.numpy.zeros(ncv, dtype=dtype)
+    initvals = [krylov_vecs, alphas, betas, 1]
+    krylov_vecs, alphas, betas, _ = jax.lax.while_loop(cond_fun, body_lanczos,
+                                                       initvals)
+    # FIXME (mganahl): if the while_loop stopps early at iteration i, alphas
+    # and betas are 0.0 at positions n >= i - 1. eigh will then wrongly give
+    # degenerate eigenvalues 0.0. JAX does currently not support
+    # dynamic slicing with variable slice sizes, so these beta values
+    # can't be truncated. Thus, if numeig >= i - 1, jitted_lanczos returns
+    # a set of spurious eigen vectors and eigen values.
+    # If algebraically small EVs are desired, one can initialize `alphas` with
+    # large positive values, thus pushing the spurious eigenvalues further
+    # away from the desired ones (similar for algebraically large EVs)
+
+    A_tridiag = jax.numpy.diag(alphas) + jax.numpy.diag(
+        betas[2:], 1) + jax.numpy.diag(jax.numpy.conj(betas[2:]), -1)
     eigvals, U = jax.numpy.linalg.eigh(A_tridiag)
```

Comment:
Might be worth flagging this for replacement with `eigh_banded` if and when Jax supports that function.

Reply:
done